Mathlib/SetTheory/Game/Pgame.lean

```diff
@@ -0,0 +1,1919 @@
+  rfl
+#align pgame.left_moves_mk Pgame.leftMoves_mk
+
+@[simp]
+theorem moveLeft_mk {xl xr xL xR} : (⟨xl, xr, xL, xR⟩ : Pgame).moveLeft = xL :=
+  rfl
+#align pgame.move_left_mk Pgame.moveLeft_mk
+
+@[simp]
+theorem rightMoves_mk {xl xr xL xR} : (⟨xl, xr, xL, xR⟩ : Pgame).RightMoves = xr :=
+  rfl
+#align pgame.right_moves_mk Pgame.rightMoves_mk
+
+@[simp]
+theorem moveRight_mk {xl xr xL xR} : (⟨xl, xr, xL, xR⟩ : Pgame).moveRight = xR :=
+  rfl
+#align pgame.move_right_mk Pgame.moveRight_mk
+
+-- TODO define this at the level of games, as well, and perhaps also for finsets of games.
+/-- Construct a pre-game from list of pre-games describing the available moves for Left and Right.
+-/
+def ofLists (L R : List Pgame.{u}) : Pgame.{u} :=
+  mk (ULift (Fin L.length)) (ULift (Fin R.length)) (fun i => L.nthLe i.down i.down.is_lt) fun j ↦
+    R.nthLe j.down j.down.prop
+#align pgame.of_lists Pgame.ofLists
+
+theorem leftMoves_ofLists (L R : List Pgame) : (ofLists L R).LeftMoves = ULift (Fin L.length) :=
+  rfl
+#align pgame.left_moves_of_lists Pgame.leftMoves_ofLists
+
+theorem rightMoves_ofLists (L R : List Pgame) : (ofLists L R).RightMoves = ULift (Fin R.length) :=
+  rfl
+#align pgame.right_moves_of_lists Pgame.rightMoves_ofLists
+
+/-- Converts a number into a left move for `ofLists`. -/
+def toOfListsLeftMoves {L R : List Pgame} : Fin L.length ≃ (ofLists L R).LeftMoves :=
+  ((Equiv.cast (leftMoves_ofLists L R).symm).trans Equiv.ulift).symm
+#align pgame.to_of_lists_left_moves Pgame.toOfListsLeftMoves
+
+/-- Converts a number into a right move for `ofLists`. -/
+def toOfListsRightMoves {L R : List Pgame} : Fin R.length ≃ (ofLists L R).RightMoves :=
+  ((Equiv.cast (rightMoves_ofLists L R).symm).trans Equiv.ulift).symm
+#align pgame.to_of_lists_right_moves Pgame.toOfListsRightMoves
+
+theorem ofLists_moveLeft {L R : List Pgame} (i : Fin L.length) :
+    (ofLists L R).moveLeft (toOfListsLeftMoves i) = L.nthLe i i.is_lt :=
+  rfl
+#align pgame.of_lists_move_left Pgame.ofLists_moveLeft
+
+@[simp]
+theorem ofLists_moveLeft' {L R : List Pgame} (i : (ofLists L R).LeftMoves) :
+    (ofLists L R).moveLeft i =
+      L.nthLe (toOfListsLeftMoves.symm i) (toOfListsLeftMoves.symm i).is_lt :=
+  rfl
+#align pgame.of_lists_move_left' Pgame.ofLists_moveLeft'
+
+theorem ofLists_moveRight {L R : List Pgame} (i : Fin R.length) :
+    (ofLists L R).moveRight (toOfListsRightMoves i) = R.nthLe i i.is_lt :=
+  rfl
+#align pgame.of_lists_move_right Pgame.ofLists_moveRight
+
+@[simp]
+theorem ofLists_moveRight' {L R : List Pgame} (i : (ofLists L R).RightMoves) :
+    (ofLists L R).moveRight i =
+      R.nthLe (toOfListsRightMoves.symm i) (toOfListsRightMoves.symm i).is_lt :=
+  rfl
+#align pgame.of_lists_move_right' Pgame.ofLists_moveRight'
+
+/-- A variant of `Pgame.recOn` expressed in terms of `Pgame.moveLeft` and `Pgame.moveRight`.
+
+Both this and `Pgame.recOn` describe Conway induction on games. -/
+-- Porting note: Code generator does not support `Pgame.recOn`
```

Comment:
How does one access the recursion principle of an inductive type in Mathlib 4?

Reply:
Same way as in lean 3. Note that the error is that the *code generator* doesn't support the recursor. If you mark it `noncomputable` then it works